From 7ee268886aac7c609b3f4855847cfbdf2e1afe71 Mon Sep 17 00:00:00 2001
From: Noah Davis <noahadvs@gmail.com>
Date: Tue, 19 May 2026 16:04:00 -0400
Subject: [PATCH 3/3] BACKPORT: Add alpha channel support with non-OpenCV stack
 blur

It didn't support the alpha channel, so it would just create black rectangles when used for shadows.

It still doesn't look the same as the OpenCV blur and it's not as fast, but at least it's not broken anymore.

BUG: 519107
(cherry picked from commit a71fc46e7512cb07f1bc7750b604811e9b175a2d)
---
 src/annotations/stackblur.cpp | 47 +++++++++++++++++++++++++++--------
 1 file changed, 36 insertions(+), 11 deletions(-)

diff --git a/src/annotations/stackblur.cpp b/src/annotations/stackblur.cpp
index d27c844afeba..ee5d4818e0f8 100644
--- a/src/annotations/stackblur.cpp
+++ b/src/annotations/stackblur.cpp
@@ -40,15 +40,15 @@ void StackBlur::blur(QImage &image, const QSize &kernelSize)
         return;
     }
 
-    const int radiusX = kernelSize.width();
-    const int radiusY = kernelSize.height();
+    const int radiusX = kernelSize.width() / 2;
+    const int radiusY = kernelSize.height() / 2;
     const int w = image.width();
     const int h = image.height();
 
     const int wm = w - 1;
     const int hm = h - 1;
 
-    std::vector<int> r(w * h), g(w * h), b(w * h);
+    std::vector<int> r(w * h), g(w * h), b(w * h), a(w * h);
     std::vector<int> vmin(std::max(w, h));
 
     // Horizontal pass with radiusX
@@ -58,13 +58,13 @@ void StackBlur::blur(QImage &image, const QSize &kernelSize)
         divsumX *= divsumX;
 
         struct Pixel {
-            int r, g, b;
+            int r, g, b, a;
         };
         std::vector<Pixel> stack(divX);
 
         for (int y = 0; y < h; y++) {
-            int rinsum = 0, ginsum = 0, binsum = 0, routsum = 0, goutsum = 0, boutsum = 0;
-            int rsum = 0, gsum = 0, bsum = 0;
+            int rinsum = 0, ginsum = 0, binsum = 0, ainsum = 0, routsum = 0, goutsum = 0, boutsum = 0, aoutsum = 0;
+            int rsum = 0, gsum = 0, bsum = 0, asum = 0;
             int yi = y * w;
 
             for (int i = -radiusX; i <= radiusX; i++) {
@@ -74,20 +74,24 @@ void StackBlur::blur(QImage &image, const QSize &kernelSize)
                 sir.r = qRed(p);
                 sir.g = qGreen(p);
                 sir.b = qBlue(p);
+                sir.a = qAlpha(p);
 
                 int rbs = radiusX + 1 - abs(i);
                 rsum += sir.r * rbs;
                 gsum += sir.g * rbs;
                 bsum += sir.b * rbs;
+                asum += sir.a * rbs;
 
                 if (i > 0) {
                     rinsum += sir.r;
                     ginsum += sir.g;
                     binsum += sir.b;
+                    ainsum += sir.a;
                 } else {
                     routsum += sir.r;
                     goutsum += sir.g;
                     boutsum += sir.b;
+                    aoutsum += sir.a;
                 }
             }
 
@@ -96,10 +100,12 @@ void StackBlur::blur(QImage &image, const QSize &kernelSize)
                 r[yi] = (rsum * mulTable[radiusX]) >> shgTable[radiusX];
                 g[yi] = (gsum * mulTable[radiusX]) >> shgTable[radiusX];
                 b[yi] = (bsum * mulTable[radiusX]) >> shgTable[radiusX];
+                a[yi] = (asum * mulTable[radiusX]) >> shgTable[radiusX];
 
                 rsum -= routsum;
                 gsum -= goutsum;
                 bsum -= boutsum;
+                asum -= aoutsum;
 
                 int stackstart = (stackpointer - radiusX + divX) % divX;
                 Pixel &sir = stack[stackstart];
@@ -107,6 +113,7 @@ void StackBlur::blur(QImage &image, const QSize &kernelSize)
                 routsum -= sir.r;
                 goutsum -= sir.g;
                 boutsum -= sir.b;
+                aoutsum -= sir.a;
 
                 if (y == 0)
                     vmin[x] = std::clamp(x + radiusX + 1, 0, wm);
@@ -115,13 +122,16 @@ void StackBlur::blur(QImage &image, const QSize &kernelSize)
                 sir.r = qRed(p);
                 sir.g = qGreen(p);
                 sir.b = qBlue(p);
+                sir.a = qAlpha(p);
 
                 rinsum += sir.r;
                 ginsum += sir.g;
                 binsum += sir.b;
+                ainsum += sir.a;
                 rsum += rinsum;
                 gsum += ginsum;
                 bsum += binsum;
+                asum += ainsum;
 
                 stackpointer = (stackpointer + 1) % divX;
 
@@ -129,9 +139,11 @@ void StackBlur::blur(QImage &image, const QSize &kernelSize)
                 routsum += sir2.r;
                 goutsum += sir2.g;
                 boutsum += sir2.b;
+                aoutsum += sir2.a;
                 rinsum -= sir2.r;
                 ginsum -= sir2.g;
                 binsum -= sir2.b;
+                ainsum -= sir2.a;
 
                 yi++;
             }
@@ -145,6 +157,7 @@ void StackBlur::blur(QImage &image, const QSize &kernelSize)
                 r[yi] = qRed(p);
                 g[yi] = qGreen(p);
                 b[yi] = qBlue(p);
+                a[yi] = qAlpha(p);
             }
         }
     }
@@ -156,13 +169,13 @@ void StackBlur::blur(QImage &image, const QSize &kernelSize)
         divsumY *= divsumY;
 
         struct Pixel {
-            int r, g, b;
+            int r, g, b, a;
         };
         std::vector<Pixel> stack(divY);
 
         for (int x = 0; x < w; x++) {
-            int rinsum = 0, ginsum = 0, binsum = 0, routsum = 0, goutsum = 0, boutsum = 0;
-            int rsum = 0, gsum = 0, bsum = 0;
+            int rinsum = 0, ginsum = 0, binsum = 0, ainsum = 0, routsum = 0, goutsum = 0, boutsum = 0, aoutsum = 0;
+            int rsum = 0, gsum = 0, bsum = 0, asum = 0;
             int yi = x;
 
             for (int i = -radiusY; i <= radiusY; i++) {
@@ -172,19 +185,23 @@ void StackBlur::blur(QImage &image, const QSize &kernelSize)
                 sir.r = r[p];
                 sir.g = g[p];
                 sir.b = b[p];
+                sir.a = a[p];
                 int rbs = radiusY + 1 - abs(i);
                 rsum += r[p] * rbs;
                 gsum += g[p] * rbs;
                 bsum += b[p] * rbs;
+                asum += a[p] * rbs;
 
                 if (i > 0) {
                     rinsum += sir.r;
                     ginsum += sir.g;
                     binsum += sir.b;
+                    ainsum += sir.a;
                 } else {
                     routsum += sir.r;
                     goutsum += sir.g;
                     boutsum += sir.b;
+                    aoutsum += sir.a;
                 }
             }
 
@@ -192,19 +209,22 @@ void StackBlur::blur(QImage &image, const QSize &kernelSize)
             for (int y = 0; y < h; y++) {
                 image.setPixel(x,
                                y,
-                               qRgb((rsum * mulTable[radiusY]) >> shgTable[radiusY],
+                               qRgba((rsum * mulTable[radiusY]) >> shgTable[radiusY],
                                     (gsum * mulTable[radiusY]) >> shgTable[radiusY],
-                                    (bsum * mulTable[radiusY]) >> shgTable[radiusY]));
+                                    (bsum * mulTable[radiusY]) >> shgTable[radiusY],
+                                    (asum * mulTable[radiusY]) >> shgTable[radiusY]));
 
                 rsum -= routsum;
                 gsum -= goutsum;
                 bsum -= boutsum;
+                asum -= aoutsum;
 
                 int stackstart = (stackpointer - radiusY + divY) % divY;
                 Pixel &sir = stack[stackstart];
                 routsum -= sir.r;
                 goutsum -= sir.g;
                 boutsum -= sir.b;
+                aoutsum -= sir.a;
 
                 if (x == 0)
                     vmin[y] = std::clamp(y + radiusY + 1, 0, hm) * w;
@@ -213,13 +233,16 @@ void StackBlur::blur(QImage &image, const QSize &kernelSize)
                 sir.r = r[p];
                 sir.g = g[p];
                 sir.b = b[p];
+                sir.a = a[p];
 
                 rinsum += sir.r;
                 ginsum += sir.g;
                 binsum += sir.b;
+                ainsum += sir.a;
                 rsum += rinsum;
                 gsum += ginsum;
                 bsum += binsum;
+                asum += ainsum;
 
                 stackpointer = (stackpointer + 1) % divY;
 
@@ -227,9 +250,11 @@ void StackBlur::blur(QImage &image, const QSize &kernelSize)
                 routsum += sir2.r;
                 goutsum += sir2.g;
                 boutsum += sir2.b;
+                aoutsum += sir2.a;
                 rinsum -= sir2.r;
                 ginsum -= sir2.g;
                 binsum -= sir2.b;
+                ainsum -= sir2.a;
 
                 yi += w;
             }
-- 
2.52.0

